docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
+## Introduction
+
+This proposal discusses adding the support for referencing an ACK custom resource
+in another custom resource. The reconciler will wait for the referenced custom
+resource to be created, and in a stable state, before using its spec or state to
+create the referencing custom resource.
+
+For example, currently if ACK customer needs to create an APIGatewayv2 integration,
+they first need to
+1. create an `Api` custom resource
+2. wait for `Api` resource creation to finish
+3. manually find the `ApiId` from `Status` of `Api` resource
+4. create an `Integration` resource with the `ApiId`
+
+Having to manually wait for the resource creation makes it impossible to create
+ACK resources using any gitops method where all the custom resource manifest
+can be managed in the source control and applied together to k8s cluster.
+
+With this approach of referenced resources, customers will be able to reference
+Api resource in Integration resource and apply both resource manifests
+together in their k8s cluster. This will allow customers to manage ACK custom
+resources using gitops methodologies. ACK controller will wait for API resource
+to get created and then use the ApiId to create Integration resource.
+
+The deletion strategy for the referenced resource is out of scope for this proposal
+right now. For Ex: If API resource is referenced in integration resource, the deletion
+of API resource will not be blocked. We plan to handle deletion strategy in next
+revision.
+
+## Scope
+
+### First Revision
+* Allow adding resource name(s) from same api-group as reference in the custom
+resource definition
+* Only the resources from same namespace will be allowed to use as reference
+* Only the fields of type '*string' or '[]*string' can be referenced from
+another resource. This satisfies the use case of referencing 'ID(s)/Name(s)/ARN(s)'
+from other resources
+
+### Future Improvements
+* Add support for custom hooks while dereferencing resources
+* Add support for `selector` for referencing the objects
+* Add support for referencing resource from different api-groups
+* Add support for not deleting a resource if it is referenced in another resource
+* Add support for finite retries and terminal conditions when resolving references
+
+## Preferred Approach
+
+### High Level Overview
+
+* Introduce new construct in generator.yaml which will indicate to code-generator
+if a new spec field with type `AWSResourceReference` should be added to the custom
+resource definition.
+
+* This new spec field will work as additional input for an existing spec field.
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
```

Comment:
> Addition of the extra spec field keeps the backward compatibility with existing custom resource definitions

Could you clarify to help me understand how it wouldn't maintain backwards compat when you don't add an extra field? if no reference is provided, then can't we default to current behavior?

It seems weird to me that we need a new field when I just want my resource's existing field substituted with this referenced value.

Reply:
> Could you clarify to help me understand how it wouldn't maintain backwards compat when you don't add an extra field?

Because then we will be changing the type of existing field to accommodate passing another resource's reference.

> if no reference is provided, then can't we default to current behavior?

This will be the behavior for preferred approach

> It seems weird to me that we need a new field when I just want my resource's existing field substituted with this referenced value.

We would need a way to pass the reference in resource spec. The new field is precisely for that.